src/main/java/com/recruitment/project/parking/parkingspaces/service/UserParkingVisitServiceImpl.java

```diff
@@ -0,0 +1,124 @@
+        userParkingVisit.setUserVehicle(userVehicle);
+        userParkingVisit.setDisabledUser(isDisabled);
+        userParkingVisit.setStartParking(LocalDateTime.now());
+        userParkingVisitRepository.save(userParkingVisit);
+    }
+
+    @Override
+    public void endUserParkingVisit(String registrationNumber) {
+        UserVehicle userVehicle = userVehicleRepository.findByRegistration(registrationNumber);
+        if (userVehicle != null) {
+            UserParkingVisit userParkingVisit = userParkingVisitRepository.findTopByUserVehicleOrderByStartParkingDesc(userVehicle);
+            userParkingVisit.setFinishParking(LocalDateTime.now());
+            Duration parkingVisitTime = Duration.between(userParkingVisit.getStartParking(), userParkingVisit.getFinishParking());
+            userParkingVisit.setCostVisit(parkingCost(parkingVisitTime.toMinutes(), userParkingVisit.isDisabledUser()));
+            userParkingVisitRepository.save(userParkingVisit);
+            logger.info(userParkingVisit.getStartParking().toString());
+        } else {
+            logger.info("Vehicle with registration number: " + registrationNumber + "did not started park");
+        }
+    }
+
+    @Override
+    public UserParkingVisit isStartingUserParkingVisit(String registrationNumber) {
+        UserVehicle userVehicle = userVehicleRepository.findByRegistration(registrationNumber);
+        if (userVehicle != null) {
+            UserParkingVisit userParkingVisit = userParkingVisitRepository.findTopByUserVehicleOrderByStartParkingDesc(userVehicle);
+            return userParkingVisit;
+        } else {
+            logger.info("Vehicle with registration number: " + registrationNumber + " did not started park");
+            return null;
+        }
+
+    }
+
+    @Override
+    public double getCostOneParkingVisit(String registrationNumber) {
+        UserVehicle userVehicle = userVehicleRepository.findByRegistration(registrationNumber);
+        if (userVehicle != null) {
+            UserParkingVisit userParkingVisit = userParkingVisitRepository.findTopByUserVehicleOrderByFinishParkingDesc(userVehicle);
+            return userParkingVisit.getCostVisit();
+        } else {
+            logger.info("Vehicle with registration number: " + registrationNumber + " did not started park");
+            return -1;
+        }
+    }
+
+    @Override
+    public double getIncomeAllParkingVisit(LocalDate dayFrom) {
+        List<UserParkingVisit> allParkingVisitFromDay = userParkingVisitRepository.findUserParkingVisitsByFinishParkingBetween(dayFrom.atStartOfDay(), dayFrom.atTime(23, 59, 59));
+        double allIncomeFromDay = 0;
+        for (UserParkingVisit userParkingVisit : allParkingVisitFromDay) {
+            allIncomeFromDay += userParkingVisit.getCostVisit();
+        }
+
+        return round(allIncomeFromDay * 100.0) / 100.0;
+    }
+
+     double parkingCost(long parkingVisitTimeInMin, boolean isDisabled) {
```

Comment:
Strasznie nieczytelna ta funkcja, masz pomysł jak zrobić to lepiej?

Reply:
1. Pomysł: Na pewno moglibyśmy wyciągnąć deklaracje zmiennej allParkingCost na początek metody i zainicjalizować zerem, a następnie zamiast 3 returnów zrobić jeden na końcu. Zastąpić else if samym if i kolejno dodawać koszty do **allParkingCost** w poszczególnych instrukcjach warunkowych, a dopiero na samym końcu zwrócić wartość całkowitego kosztu parkowania .
2. Pomysł: Moglibyśmy wyciągnąć if (isDisable) na początek, zwiększyło by to czytelność, ale kosztem redundancji kodu.

Reply:
Można by też użyć wzorca jakiegoś, masz pomysł jakiego?

Reply:
Tutaj użyłbym fabryki, wtedy można by było używać metody do obliczenia kosztu bez parametru, a w przypadku pojawienia się nowej taryfy (trzeciej) faktycznie łatwiej byłoby to rozszerzyć (open/closed principle).

Reply:
Wzorzec strategia